src/notebooks/controllers/controllerLoader.ts

```diff
@@ -0,0 +1,260 @@
+import { createActiveInterpreterController } from './helpers';
+import { IControllerLoader, IControllerRegistration } from './types';
+
+// Even after shutting down a kernel, the server API still returns the old information.
+// Re-query after 2 seconds to ensure we don't get stale information.
+const REMOTE_KERNEL_REFRESH_INTERVAL = 2_000;
+
+/**
+ * This class finds and creates notebook controllers.
+ */
+@injectable()
+export class ControllerLoader implements IControllerLoader, IExtensionSyncActivationService {
+    private get isLocalLaunch(): boolean {
+        return this.serverConnectionType.isLocalLaunch;
+    }
+    private wasPythonInstalledWhenFetchingControllers?: boolean;
+    private refreshedEmitter = new vscode.EventEmitter<void>();
+    // Promise to resolve when we have loaded our controllers
+    private controllersPromise: Promise<void>;
+    constructor(
+        @inject(IVSCodeNotebook) private readonly notebook: IVSCodeNotebook,
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        @inject(IExtensions) private readonly extensions: IExtensions,
+        @inject(IKernelFinder) private readonly kernelFinder: IKernelFinder,
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider,
+        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker,
+        @inject(IInterpreterService) private readonly interpreters: IInterpreterService,
+        @inject(IJupyterServerUriStorage) private readonly serverUriStorage: IJupyterServerUriStorage,
+        @inject(ServerConnectionType) private readonly serverConnectionType: ServerConnectionType,
+        @inject(IControllerRegistration) private readonly registration: IControllerRegistration
+    ) {
+        let timer: NodeJS.Timeout | number | undefined;
+        this.interpreters.onDidChangeInterpreters(
+            () => {
+                if (timer) {
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    clearTimeout(timer as any);
+                }
+                timer = setTimeout(
+                    () =>
+                        this.loadControllers(true).catch((ex) =>
+                            traceError('Failed to re-query python kernels after changes to list of interpreters', ex)
+                        ),
+                    // This hacky solution should be removed in favor of https://github.com/microsoft/vscode-jupyter/issues/7583
+                    // as a proper fix for https://github.com/microsoft/vscode-jupyter/issues/5319
+                    1_000
+                );
+            },
+            this,
+            this.disposables
+        );
+
+        // Make sure to reload whenever we do something that changes state
+        const forceLoad = () => this.loadControllers(true);
+        this.serverUriStorage.onDidChangeUri(forceLoad, this, this.disposables);
```

Comment:
Would be best to move all of these to the `activate` method, else seems inconsistent to have some event handlers in the ctro and others in `activate`

Reply:
I disagree. At least the initial load has to happen in the ctor. 

Otherwise another IExtensionSingleActivationService could take a dependency on the controllerLoader and have the loadpromise not even start yet.